### PR TITLE
CI: reduce running time for PHP unit testing jobs

### PR DESCRIPTION
**Pull request by kalessil**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Reduce execution times for PHPUnit-related jobs (5 currently) by skipping the build step (mostly about building assets). Build times vary, but on average, they're around 2.5 minutes (?).

Full scope:
- PHPUnit and API tests: skipping building assets (no interactions with UI)
- Linting: skipping building assets (not needed to perform linting)

### How to test the changes in this Pull Request:

See [here](https://github.com/woocommerce/woocommerce/actions/runs/9793406791/job/27041437354?pr=49135) and note `Build project` is skipped in `PHP *` jobs, which are green.

### CODE REVIEW

File: plugins/woocommerce/composer.lock [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
     "_readme": [
-        "This file locks the dependencies of your project to a known state",
+        "This file locks the dependencies of your project to a known state.",
```
Comment (kalessil) [inline plugins/woocommerce/composer.lock]: To be reverted before merge.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -118,15 +118,13 @@ jobs:
         id: 'install-monorepo'
         with:
           install: '${{ matrix.projectName }}...'
-          build: 'false'
           pull-playwright-cache: ${{ matrix.testEnv.shouldCreate && matrix.testType == 'e2e' }}
 
       - uses: './.github/actions/setup-woocommerce-monorepo'
-        if: ${{ github.ref_type != 'tag' }}
+        if: ${{ github.ref_type != 'tag' && matrix.testType != 'unit:php' }}
```
Comment (kalessil) [inline .github/workflows/ci.yml]: Should we try applying the same approach for `API` tests?

**Comment by adimoldovan** [comment]
@kalessil You'll need to add the testType to all the package.json files in `packages/js/*` and most of them will need a build.

How do you feel about adding a `build` option to the job config object instead of using the test type? We could also use it for lint jobs.
I tried skipping the build for lint jobs in https://github.com/woocommerce/woocommerce/pull/48234 but had to revert it because woocommerce-blocks need still the build step.

**Comment by kalessil** [comment]
@adimoldovan 

> You'll need to add the testType to all the package.json files in packages/js/* and most of them will need a build.

IIUC, the type fallbacks to 'unit:js' here: https://github.com/woocommerce/woocommerce/pull/49135/files#diff-475b76cd4757faceb26cf6f193e4494f3262119dab0e682ebd6876d430d89bfcR409 (hope I understand your feedback correctly).

> How do you feel about adding a build option to the job config object instead of using the test type? We could also use it for lint jobs.

Sounds like a solid approach to me. Do we have a technical option to move CI-related configs into dedicated files (that would be a great refactoring combo)? As a note, I'd prefer to work on that in a follow-up PR.

> I tried skipping the build for lint jobs in https://github.com/woocommerce/woocommerce/pull/48234 but had to revert it because woocommerce-blocks need still the build step.

Gotcha, will keep in mind for the build option implementation.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (kalessil) [reply .github/workflows/ci.yml]: API tests are passing as well with the same approach.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -141,7 +139,11 @@ jobs:
         id: 'prepare-test-environment'
         if: ${{ matrix.testEnv.shouldCreate }}
         env: ${{ matrix.testEnv.envVars }}
-        run: 'pnpm --filter="${{ matrix.projectName }}" ${{ matrix.testEnv.start }}'
+        run: |
+          if [[ '${{ matrix.testType }}' == 'unit:php' || '${{ matrix.testType }}' == 'api' ]]; then
```
Comment (kalessil) [inline .github/workflows/ci.yml]: TBD: in this PR rework in favour of 'build_type' input for the `setup-woocommerce-monorepo` action.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -118,15 +118,13 @@ jobs:
         id: 'install-monorepo'
         with:
           install: '${{ matrix.projectName }}...'
-          build: 'false'
           pull-playwright-cache: ${{ matrix.testEnv.shouldCreate && matrix.testType == 'e2e' }}
 
       - uses: './.github/actions/setup-woocommerce-monorepo'
-        if: ${{ github.ref_type != 'tag' }}
+        if: ${{ github.ref_type != 'tag' && matrix.testType != 'unit:php' && matrix.testType != 'api' }}
```
Comment (kalessil) [inline .github/workflows/ci.yml]: TBD: merge with the preceding step.

File: tools/monorepo/check-changelogger-use.php [hunk 5]
```diff
@@ -193,6 +193,11 @@ function debug() {
 		}
 		continue;
 	}
+	// Ignore dot-files: those are development related, and it makes no sense to create a changelog entry for them.
+	if ( '.' === basename( $line )[0] ) {
```
Comment (kalessil) [inline tools/monorepo/check-changelogger-use.php:197]: Adding as eslint config changes should not require changelog entry.

**Comment by adimoldovan** [comment]
@kalessil 
> IIUC, the type fallbacks to 'unit:js' here: [#49135 (files)](https://github.com/woocommerce/woocommerce/pull/49135/files#diff-475b76cd4757faceb26cf6f193e4494f3262119dab0e682ebd6876d430d89bfcR409) (hope I understand your feedback correctly).

Got it, I missed that part. Although, falling back to unit:js is not ideal IMO. Maybe keep the generic "unit" as fallback? Would that complicate things too much? 🤔 

> Sounds like a solid approach to me. Do we have a technical option to move CI-related configs into dedicated files (that would be a great refactoring combo)? As a note, I'd prefer to work on that in a follow-up PR.

A follow-up PR sounds good to me.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (adimoldovan) [reply .github/workflows/ci.yml]: API should required the same build as e2e tests.

Comment (kalessil) [reply .github/workflows/ci.yml]: API tests are REST API tests, correct @adimoldovan ?

**Comment by kalessil** [comment]
> Although, falling back to unit:js is not ideal IMO. Maybe keep the generic "unit" as fallback?

Yep, I'll revert renaming `unit` to `unit:js` to minimize changes.

> A follow-up PR sounds good to me.

Perfect, thank you!

**Comment by kalessil** [comment]
Splitting the PR into two separate ones to ensure changes are better scoped and smaller shipping chunks:
- https://github.com/woocommerce/woocommerce/pull/49193
- https://github.com/woocommerce/woocommerce/pull/49194

**Pull request closed without merging**